File: src/logic/regions.js
// EPIC-01 地域を登録できる。
import { loadState, saveState, nextId, nowIso } from "./db.js"

const ZIPCLOUD_URL = "https://zipcloud.ibsnet.co.jp/api/search"

function normalizeZipcode(zipcode) {
  const digits = zipcode.replace(/-/g, "").trim()
  if (!/^\d{7}$/.test(digits)) {
    throw new Error(`郵便番号の形式が不正です: ${zipcode}`)
  }
  return digits
}

// zipcloud APIはCORSに対応していないため、JSONPで呼び出す。
function requestJsonp(url) {
  return new Promise((resolve, reject) => {
    const callbackName = `__zipcloudCallback${Date.now()}${Math.floor(Math.random() * 1e6)}`
    const script = document.createElement("script")
    const cleanup = () => {
      delete window[callbackName]
      script.remove()
    }
    window[callbackName] = (body) => {
      cleanup()
      resolve(body)
    }
    script.onerror = () => {
      cleanup()
      reject(new Error("郵便番号検索APIの呼び出しに失敗しました"))
    }
    script.src = `${url}${url.includes("?") ? "&" : "?"}callback=${callbackName}`
    document.head.appendChild(script)
  })
}

export async function lookupMunicipality(zipcode) {
  const digits = normalizeZipcode(zipcode)
  const body = await requestJsonp(`${ZIPCLOUD_URL}?zipcode=${digits}`)
  const results = body.results
  if (!results || results.length === 0) {
    throw new Error(`郵便番号に該当する住所が見つかりません: ${zipcode}`)
  }
  const top = results[0]
  return {
    prefecture: top.address1,
    city: top.address2,
    town: top.address3,
  }
}

export async function registerRegion(zipcode) {
  const municipality = await lookupMunicipality(zipcode)
  const state = loadState()
  const now = nowIso()
  const region = {
    id: nextId(state, "regions"),
    zipcode: normalizeZipcode(zipcode),
    prefecture: municipality.prefecture,
    city: municipality.city,
    town: municipality.town,
    created_at: now,
    updated_at: now,
  }
  state.regions.push(region)
  saveState(state)
  return region
}

export function getRegion(regionId, state = loadState()) {
  const region = state.regions.find((r) => r.id === regionId)
  if (!region) {
    throw new Error(`地域が見つかりません: id=${regionId}`)
  }
  return region
}

export async function updateRegion(regionId, zipcode) {
  const state = loadState()
  const region = getRegion(regionId, state)
  const municipality = await lookupMunicipality(zipcode)
  const now = nowIso()
  region.zipcode = normalizeZipcode(zipcode)
  region.prefecture = municipality.prefecture
  region.city = municipality.city
  region.town = municipality.town
  region.updated_at = now
  saveState(state)
  return region
}

export function deleteRegion(regionId) {
  const state = loadState()
  getRegion(regionId, state) // 存在確認
  state.regions = state.regions.filter((r) => r.id !== regionId)
  saveState(state)
}

export function listRegions() {
  return loadState().regions
}

File: app/regions.py
"""EPIC-01 地域を登録できる。"""
import json
import re
import sqlite3
import urllib.request
from datetime import datetime, timezone

ZIPCLOUD_URL = "https://zipcloud.ibsnet.co.jp/api/search?zipcode={zipcode}"


def _normalize_zipcode(zipcode: str) -> str:
    digits = zipcode.replace("-", "").strip()
    if not re.fullmatch(r"\d{7}", digits):
        raise ValueError(f"郵便番号の形式が不正です: {zipcode}")
    return digits


def lookup_municipality(zipcode: str) -> dict:
    """郵便番号から都道府県・市区町村・町域を特定する（zipcloud APIを利用）。"""
    digits = _normalize_zipcode(zipcode)
    with urllib.request.urlopen(ZIPCLOUD_URL.format(zipcode=digits), timeout=10) as res:
        body = json.loads(res.read())
    results = body.get("results")
    if not results:
        raise ValueError(f"郵便番号に該当する住所が見つかりません: {zipcode}")
    top = results[0]
    return {
        "prefecture": top["address1"],
        "city": top["address2"],
        "town": top["address3"],
    }


def register_region(conn: sqlite3.Connection, zipcode: str) -> dict:
    municipality = lookup_municipality(zipcode)
    now = datetime.now(timezone.utc).isoformat()
    cur = conn.execute(
        "INSERT INTO regions (zipcode, prefecture, city, town, created_at, updated_at) "
        "VALUES (?, ?, ?, ?, ?, ?)",
        (_normalize_zipcode(zipcode), municipality["prefecture"], municipality["city"],
         municipality["town"], now, now),
    )
    conn.commit()
    return get_region(conn, cur.lastrowid)


def get_region(conn: sqlite3.Connection, region_id: int) -> dict:
    row = conn.execute("SELECT * FROM regions WHERE id = ?", (region_id,)).fetchone()
    if row is None:
        raise ValueError(f"地域が見つかりません: id={region_id}")
    return dict(row)


def update_region(conn: sqlite3.Connection, region_id: int, zipcode: str) -> dict:
    get_region(conn, region_id)  # 存在確認
    municipality = lookup_municipality(zipcode)
    now = datetime.now(timezone.utc).isoformat()
    conn.execute(
        "UPDATE regions SET zipcode = ?, prefecture = ?, city = ?, town = ?, updated_at = ? "
        "WHERE id = ?",
        (_normalize_zipcode(zipcode), municipality["prefecture"], municipality["city"],
         municipality["town"], now, region_id),
    )
    conn.commit()
    return get_region(conn, region_id)


def delete_region(conn: sqlite3.Connection, region_id: int) -> None:
    get_region(conn, region_id)  # 存在確認
    conn.execute("DELETE FROM regions WHERE id = ?", (region_id,))
    conn.commit()
